src/walk.rs

```diff
@@ -362,6 +362,17 @@ fn spawn_senders(
                 return ignore::WalkState::Continue;
             }
 
+            // when pruning, don't descend into matching dirs
+            // note that prune requires pattern
+            let walk_action = if config.prune
+                && pattern.as_str().len() != 0
```

Comment:
I don't really understand why this is necessary?

Reply:
I added it so that the prune is only evaluated if there is a pattern specified, thinking that the prune option should only be used with a pattern. (and I added a test case for this). But, I suppose it could be allowed and then the expected behavior would be that no directories would be descended into. What do you think/prefer?

Reply:
Hm, "not having a pattern" does not only refer to cases where `--type` is used, right? It could also be a unconstrained search, a `--size` search or a `--changed-{within,before}` search.

I agree that `--prune` maybe doesn't make sense in these cases, but we should take a closer look, I think. We could use the `find` behavior as a guideline.

If two options in combination do not make sense, we should also think about disallowing that via `.conflicts_with`.

Reply:
> Hm, "not having a pattern" does not only refer to cases where `--type` is used, right? It could also be a unconstrained search, a `--size` search or a `--changed-{within,before}` search.

Good point! I hadn't considered that...

> I agree that `--prune` maybe doesn't make sense in these cases, but we should take a closer look, I think. We could use the `find` behavior as a guideline.

Lets see. If we are using `find` as an example, it seems like the `-prune` option actually takes an argument on what to prune. 
(I found this explanation helpful: https://stackoverflow.com/questions/1489277/how-to-use-prune-option-of-find-in-sh)

Currently this implementation uses the `<pattern>` as what to also prune. But I suppose we could make `--prune` take an argument, stating what to prune. In this case, the user has more control over what to prune. for example:

`$  fd foo --prune=bar`

, will look for matches with foo but not descend into dirs matching bar.

This would eliminate the need to do the check for pattern on line 369.
What do you think?
Yet it almost sounds like the functionality of `--exclude <pattern>`  but specific to directories. hmmmm... are they too similar?

> If two options in combination do not make sense, we should also think about disallowing that via `.conflicts_with`.

That is good to know about!
After reading your comment, and looking at the find documentation
http://man7.org/linux/man-pages/man1/find.1.html

- The `find -prune` seems to conflict with `-depth`, so that could be something to consider, to mark conflicts _with on `--max-depth`?

- I'm not convinced we'd need to mark prune as conflicts with --size or --changed-{} , if we added an argument to it as I proposed above. Someone could want to see files of a certain size but omitting some directories, maybe? But if we don't go with adding an argument to prune, marking as conflicts with --size and --changed-{} might be a better way to enforce the need for a <pattern> to prune on.

Reply:
> Yet it almost sounds like the functionality of --exclude <pattern> but specific to directories. hmmmm... are they too similar?

FWIW, the -E option invokes ripgreps gitignore style excludes. To ignore a directory `-E foo/` should work. Eg.

```
[squirt ~/tmp/tests]
$ tree -p
.
├── [drwxr-xr-x]  bar
│   └── [drwxr-xr-x]  lower_dir
│       └── [-rw-r--r--]  bar
└── [drwxr-xr-x]  foo
    └── [drwxr-xr-x]  lower_dir
        └── [-rw-r--r--]  bar

4 directories, 2 files
[squirt ~/tmp/tests]
$ fd -E bar/
foo
foo/lower_dir
foo/lower_dir/bar
```